// mat-select-search/esm2020/lib/mat-select-search.module.mjs
import { NgModule } from '@angular/core';
import { MatSelectSearchComponent } from './mat-select-search.component';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatDividerModule } from '@angular/material/divider';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class MatSelectSearchModule {
}
MatSelectSearchModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: MatSelectSearchModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MatSelectSearchModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.5", ngImport: i0, type: MatSelectSearchModule, declarations: [MatSelectSearchComponent], imports: [MatProgressSpinnerModule,
        MatDividerModule,
        CommonModule], exports: [MatSelectSearchComponent] });
MatSelectSearchModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: MatSelectSearchModule, imports: [MatProgressSpinnerModule,
        MatDividerModule,
        CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: MatSelectSearchModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MatSelectSearchComponent
                    ],
                    imports: [
                        MatProgressSpinnerModule,
                        MatDividerModule,
                        CommonModule
                    ],
                    exports: [
                        MatSelectSearchComponent
                    ]
                }]
        }] });
